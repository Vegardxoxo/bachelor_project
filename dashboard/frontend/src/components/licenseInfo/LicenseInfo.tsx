import React, {useEffect, useState} from 'react';
import {useLocation, useParams} from 'react-router-dom';
import {fetchInfoBoxLicense, fetchSoftwareUsedInOrg} from '../../api/calls';
import SoftwareSearchBar from '../search/SoftwareSeachBar';
import {OwnOrgData} from "../../Interfaces";
import {Box, Grid, Stack} from '@mui/material';
import OwnTable from "../licensepool/ownTable/OwnTable";
import Pagination from '@mui/material/Pagination';
import ActiveLastBreadcrumb from '../dashboard/ActivateLastBreadcrumb/ActivateLastBreadcrumb';
import MuiLoadingSpinner from '../spinner/MuiLoadingSpinner';
import {useRecoilValue} from "recoil";
import {refreshTableAtom, userAtom} from "../../globalVariables/variables";
import Checkbox from "@mui/material/Checkbox";
import FormControlLabel from "@mui/material/FormControlLabel";
import {Typography} from 'antd';

/**
 * This component is responsible for rendering the license information for the unit.
 */
const LicenseInfo: React.FC = () => {
    const storedOrganization: string | null = JSON.parse(localStorage.getItem('organization') ?? 'null');
    const {title} = useParams();
    const useQuery = () => {
        return new URLSearchParams(useLocation().search);
    };
    const query = useQuery();
    const searchTermFromUrl = query.get("searchTerm");
    const [data, setData] = useState<OwnOrgData[]>([]);
    const [searchTerm, setSearchTerm] = useState<string>(searchTermFromUrl || "");
    const [orgSoftware, setOrgSoftware] = useState<string[]>([]);
    const [status, setStatus] = useState<string | null>(null);
    const [currentPage, setCurrentPage] = useState<number>(1);
    const [count, setCount] = useState<number>(0);
    const [sortBy, setSortBy] = useState<string>('application_name')
    const [loaded, setLoaded] = React.useState(false);
    const refreshTable = useRecoilValue(refreshTableAtom)
    const [checked, setChecked] = useState<boolean>(false);
    const userData = useRecoilValue(userAtom)
    const email = checked ? userData.primary_user_email : ''


    useEffect(() => {
        switch (title) {
            case 'Totale Lisenser':
                setStatus('active')
                break;
            case 'Ubrukte Lisenser':
                setStatus('unused')
                break;
            case 'Ledige Lisenser':
                setStatus('available')
                break;
            default:
                setStatus(null);
                break;
        }
        // Fetches distinct software names.
        const fetchSoftwareNames = async () => {
            if (status && storedOrganization) {
                try {
                    const data: string[] | undefined = await fetchSoftwareUsedInOrg(status, 'false', storedOrganization, email);
                    if (data !== undefined) {
                        setOrgSoftware(data);
                    }
                } catch (error) {
                    console.error('Error fetching software names:', error);
                }
            }
        };


        fetchSoftwareNames();
    }, [status, refreshTable, checked]);

    useEffect(() => {
        fetchData();
    }, [searchTerm, currentPage, status, sortBy, refreshTable, checked]);

    // Fetches license information
    const fetchData = async () => {
        if (status && storedOrganization) {
            try {
                const data = await fetchInfoBoxLicense(currentPage, status as string,
                    sortBy as string, storedOrganization as string, searchTerm, email);
                data?.results && setData(data.results);
                data?.count && setCount(data.count);
                setLoaded(true);
            } catch (error) {
                console.error('Error fetching license data:', error);
            }
        }
    };

    // Function that gets input from the searchBar component.
    const handleChange = (term: string) => {
        setSearchTerm(term);
        setCurrentPage(1);
    }

    // Function that gets input from the Checkbox component
    const handleCheckboxChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setChecked(event.target.checked);
        setCurrentPage(1);
    }


    const handlePageChange = async (event: React.ChangeEvent<unknown>, value: number) => {
        setLoaded(false)
        setCurrentPage(value);
    };

    const handleSorting = async (sortBy: string) => {
        setLoaded(false)
        setSortBy(sortBy);
        fetchData()
    };


    return (
        <>
            <div>
                <Grid>
                    <ActiveLastBreadcrumb/>
                </Grid>
                {loaded ? (
                    <Box id={'licensepool_container'}
                         style={{
                             display: 'flex',
                             justifyContent: 'center',
                             alignContent: "center",
                             marginTop: "20px",
                             fontFamily: 'Source Sans Pro,sans-serif'
                         }}>
                        <Grid container className='license_pool' marginLeft={"-2%"}>
                            <Grid container justifyContent="center" alignItems="center" className={'license_table'}
                                  width={"100%"}>
                                <Stack direction={"column"} spacing={1} width={"70%"} marginBottom={"10px"}>
                                    <h2 style={{fontFamily: 'Source Sans Pro,sans-serif'}}>
                                        {title} i {storedOrganization}
                                    </h2>
                                    <Stack direction={'row'} spacing={5} width={"95%"} marginBottom={"30px"}
                                           alignItems="center"
                                           marginTop={"10px"}
                                           paddingBottom={"20px"}>
                                        <SoftwareSearchBar data={orgSoftware} setSelectedSoftware={handleChange}
                                                           initialValue={searchTerm}/>

                                        <FormControlLabel
                                            control={
                                                <Checkbox
                                                    checked={checked}
                                                    onChange={handleCheckboxChange}
                                                    inputProps={{'aria-label': 'controlled'}}
                                                />
                                            }

                                            label={<Typography
                                                style={{fontFamily: 'Source Sans Pro,sans-serif', fontSize: '13pt'}}>Vis
                                                mine lisenser</Typography>}
                                        />
                                    </Stack>
                                    <OwnTable data={data} handleSorting={handleSorting}/>
                                    <Pagination
                                        count={Math.ceil(count / 10)}
                                        page={currentPage}
                                        onChange={handlePageChange}
                                        color={"primary"}
                                        style={{marginTop: '1rem'}}
                                    />
                                </Stack>
                            </Grid>
                        </Grid>
                    </Box>) : (<MuiLoadingSpinner/>)}
            </div>
        </>)
};

export default LicenseInfo;