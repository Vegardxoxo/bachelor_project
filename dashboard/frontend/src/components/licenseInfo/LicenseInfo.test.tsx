import '@testing-library/jest-dom/extend-expect';
import {render, screen} from '@testing-library/react';
import {MemoryRouter, Route, Routes} from 'react-router-dom';
import {RecoilRoot} from 'recoil';
import LicenseInfo from './LicenseInfo';
import renderer from "react-test-renderer";
import React from "react";

const mockOrgs = [[
    "Adobe Acrobat DC Professional",
]];

const mockData = [{
    "application_name": "Adobe Acrobat DC Professional",
    "primary_user_full_name": "Bertil NedregÃ¥rd",
    "computer_name": "TK010027481557",
    "details": [
        {
            "id": 444282,
            "last_used": "2023-01-20"
        }
    ],
    "status": "Aktiv"
}];

jest.mock('../../api/calls', () => ({
    fetchSoftwareUsedInOrg: jest.fn(() => Promise.resolve(mockOrgs)),
    fetchInfoBoxLicense: jest.fn(() => Promise.resolve(mockData)),
}));
jest.mock('@mui/material/Pagination', () => {
    const MockPagination = () => <div data-testid="mock-pagination"/>;
    MockPagination.displayName = 'Pagination';
    return MockPagination;
});

describe('LicenseInfo page', () => {

    beforeEach(() => {
        localStorage.setItem('organization', JSON.stringify('IT-tjenesten'));
    });

    afterEach(() => {
        localStorage.clear();
    });
    it('renders the total license page', async () => {
        render(
            <MemoryRouter initialEntries={['/Totale Lisenser']}>
                <RecoilRoot>
                    <Routes>
                        <Route path='/:title' element={<LicenseInfo/>}/>
                    </Routes>
                </RecoilRoot>
            </MemoryRouter>
        );
        expect(await screen.findByText('Totale Lisenser i IT-tjenesten')).toBeInTheDocument();
    });

    it('matches snapshot', async () => {
        const testRenderer = renderer.create(
            <MemoryRouter initialEntries={['/Ubrukte Lisenser']}>
                <RecoilRoot>
                    <Routes>
                        <Route path='/:title' element={<LicenseInfo/>}/>
                    </Routes>
                </RecoilRoot>
            </MemoryRouter>
        );
        expect(testRenderer.toJSON()).toMatchSnapshot();
    })
    it('renders the unused license page', async () => {
        render(
            <MemoryRouter initialEntries={['/Ubrukte Lisenser']}>
                <RecoilRoot>
                    <Routes>
                        <Route path='/:title' element={<LicenseInfo/>}/>
                    </Routes>
                </RecoilRoot>
            </MemoryRouter>
        );

        expect(await screen.findByText('Ubrukte Lisenser i IT-tjenesten')).toBeInTheDocument();
    });
    it('renders the correct license data', async () => {
        render(
            <MemoryRouter initialEntries={['/Ledige Lisenser']}>
                <RecoilRoot>
                    <Routes>
                        <Route path='/:title' element={<LicenseInfo/>}/>
                    </Routes>
                </RecoilRoot>
            </MemoryRouter>
        );
        expect(await screen.findByText('Ledige Lisenser i IT-tjenesten')).toBeInTheDocument();
    });
});
