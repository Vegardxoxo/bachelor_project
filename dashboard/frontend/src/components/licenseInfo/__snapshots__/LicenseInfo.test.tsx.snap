// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LicenseInfo page matches snapshot 1`] = `
<div>
  <div
    className="MuiGrid-root css-vj1n65-MuiGrid-root"
  >
    <nav
      aria-label="breadcrumb"
      className="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-63vwl1-MuiTypography-root-MuiBreadcrumbs-root"
      id="breadcrumbs"
    >
      <ol
        className="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
      >
        <li
          className="MuiBreadcrumbs-li"
        >
          <a
            className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineHover css-le90w3-MuiTypography-root-MuiLink-root"
            href="/"
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
          >
            Dashboard
          </a>
        </li>
        <li
          aria-hidden={true}
          className="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
        >
          /
        </li>
        <li
          className="MuiBreadcrumbs-li"
        >
          <p
            className="MuiTypography-root MuiTypography-body1 css-1gqyz35-MuiTypography-root"
          />
        </li>
      </ol>
    </nav>
  </div>
  <div
    className="MuiBox-root css-1lb6n3v"
  >
    <span
      className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
      role="progressbar"
      style={
        Object {
          "height": 80,
          "width": 80,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </span>
  </div>
</div>
`;
